crates/rolldown_utils/src/hash_placeholder.rs

```diff
@@ -30,16 +35,19 @@ fn is_hash_placeholder(s: &str) -> bool {
   }
 
   // All characters must be alphanumeric or _ or $
-  content.chars().all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '$')
+  content.bytes().all(|c| c.is_ascii_alphanumeric() || c == b'_' || c == b'$')
 }
 
 /// Finds all hash placeholders in a string and returns their positions and values
-fn find_hash_placeholders(s: &str) -> Vec<(usize, usize, &str)> {
+pub fn find_hash_placeholders<'a>(
+  s: &'a str,
+  finder: &Finder<'static>,
+) -> Vec<(usize, usize, &'a str)> {
   // pre-allocate, the max number of placeholders is s.len() / 2
   let mut results = Vec::with_capacity(s.len() / 2);
   let mut start = 0;
 
-  while let Some(left_pos) = s[start..].find(HASH_PLACEHOLDER_LEFT) {
+  while let Some(left_pos) = finder.find(&s.as_bytes()[start..]) {
     let left_pos = start + left_pos;
     if let Some(right_pos) = s[left_pos..].find(HASH_PLACEHOLDER_RIGHT) {
```

Comment:
Searching `HASH_PLACEHOLDER_RIGHT` probably don't need `memchr` because it's only a few characters away.